yarn-project/noir-contracts/src/contracts/uniswap_contract/src/main.nr

```diff
@@ -1,133 +1,163 @@
-            hash_bytes[i + 196] = secretHash_bytes[i];
-            hash_bytes[i + 228] = deadlineForL1ToL2Message_bytes[i];
-            hash_bytes[i + 260] = canceller_bytes[i];
-            hash_bytes[i + 292] = callerOnL1_bytes[i];
+    // Since the token bridge burns funds on behalf of this contract, this contract has to tell the token contract if the signature is valid
+    // implementation is similar to how account contracts validate public approvals.
+    // if valid, it returns the IS_VALID selector which is expected by token contract
+    #[aztec(public)]
+    fn is_valid_public(message_hash: Field) -> Field {
+        let value = storage.approved_action.at(message_hash).read();
+        if (value){
+            IS_VALID_SELECTOR
+        } else {
+            0
         }
+    }
 
-        let content_sha256 = dep::std::hash::sha256(hash_bytes);
-
-        // // Convert the content_sha256 to a field element
-        let mut v = 1;
-        let mut high = 0 as Field;
-        let mut low = 0 as Field;
+    // This helper method approves the bridge to burn this contract's funds and exits the input asset to L1
+    // Assumes contract already has funds
+    // Note that private can't read public return values so created an internal public that handles everything
+    // this method is used for both private and public swaps.
+    #[aztec(public)]
+    internal fn _approve_bridge_and_exit_input_asset_to_L1(
+        token_bridge: AztecAddress, 
+        amount: Field,
+    ) {
+        let token = TokenBridge::at(token_bridge.address).token(context);
+        
+        // approve bridge to burn this contract's funds (required when exiting on L1, as it burns funds on L2):
+        let nonce_for_burn_approval = storage.nonce_for_burn_approval.read();
+        let selector = compute_selector("burn_public((Field),Field,Field)");
+        let message_hash = compute_message_hash([token_bridge.address, token, selector, context.this_address(), amount, nonce_for_burn_approval]);
+        storage.approved_action.at(message_hash).write(true);
 
-        for i in 0..16 {
-            high = high + (content_sha256[15 - i] as Field) * v;
-            low = low + (content_sha256[16 + 15 - i] as Field) * v;
-            v = v * 256;
-        }
+        // increment nonce_for_burn_approval so it won't be used again
+        storage.nonce_for_burn_approval.write(nonce_for_burn_approval + 1);
 
-        // Abuse that a % p + b % p = (a + b) % p and that low < p
-        let content_hash = low + high * v;
+        // Exit to L1 Uniswap Portal !
+        TokenBridge::at(token_bridge.address).exit_to_l1_public(
+            context,
+            context.this_portal_address(),
+            amount,
+            context.this_portal_address(),
+            nonce_for_burn_approval,
+        );
+    }
 
-        content_hash
+    #[aztec(public)]
+    internal fn _assert_token_is_same(token: Field, token_bridge: Field) {
```

Comment:
As mentioned earlier, this could happen in `_approve_bridge_and_exit_input_asset_to_L1`